### PR TITLE
FIX: add const with fused type memory view in evaluate_spline

### PR DESCRIPTION
**Pull request by OmarManzoor**
#### Reference issue

This issue was identified in https://github.com/scikit-learn/scikit-learn/pull/25624

#### What does this implement/fix?

- Adds const with the fused type memory view of the parameter `c` in the evaluate_spline function.
- Adds a test for testing BSpline with read only parameters.

#### Additional information

CC: @jjerphan

### CODE REVIEW

File: scipy/interpolate/tests/test_bsplines.py [hunk 1]
```diff
@@ -592,6 +592,18 @@ def test_from_power_basis_exmp(self):
                                         bc_type='natural')
         assert_allclose(bspl.c, [1, 1, 1, 1, 1, 1, 1], atol=1e-15)
 
+    def test_read_only(self):
+        t = np.array([0, 1])
```
Comment (jjerphan) [inline scipy/interpolate/tests/test_bsplines.py:596]:
```suggestion
    def test_read_only(self):
        # BSpline must work on read-only knots and coefficients.
        t = np.array([0, 1])
```

**Comment by jjerphan** [comment]
I have opened https://github.com/scipy/scipy/pull/18154 which is a per-requisite for this PR.

**Comment by ev-br** [comment]
The fix LGTM. CI errors seem unrelated, best rebase on main now that gh-18154 is in however. 

There also several other instances where `const double_or_complex` would be nice to have. Those can be done either here or as a follow-up though.

**Comment by OmarManzoor** [comment]
> The fix LGTM. CI errors seem unrelated, best rebase on main now that [gh-18154](https://github.com/scipy/scipy/pull/18154) is in however.
> 
> There also several other instances where `const double_or_complex` would be nice to have. Those can be done either here or as a follow-up though.

Thank you for the review. How about changing the other instances in a follow up PR?

**Comment by ev-br** [comment]
Thanks @OmarManzoor! Keep them coming :-)

**Comment by jjerphan** [comment]
Thank you, @ev-br. 🙂

**Pull request merged**